Add Race timer tutorial

CommanderFoo opened the pull request:
# Description

Race timer tutorial first draft completed.

Unsure on total time it will take.  Needs to be added to the top when I work that out.

PopThosePringles committed:
Race timer tutorial

PopThosePringles committed:
Loop added to video tag.

Since these are quite small it should be ok to loop these to be consistent with the rest of the docs.  Originally I removed the attribute but in doing so videos would reach their end point while not in the viewport and it may not be obvious to the use it's a video rather than a static image.

PopThosePringles committed:
Fixed typos

PopThosePringles committed:
Removed double space (old habit)

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+If you enter **Play** mode you will notice the spawn position isn't in an ideal location for any player that joins the game. Move the **Spawn Point** so that it's before the starting position on the track.
+
+!!! tip
+    Pressing ++V++ will toggle the visibility of Gizmos. With visibility on, it's easy to see where to move objects such as the **Spawn Point**.
```
Change: "With visibility on"
to: "With gizmo visibility on"

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+2. Create a **Client Context** inside the **Scripts** folder.
+3. Create a **Server Context** inside the **Scripts** folder.
+
+### Client Context vs Server Context
```
This section will be better suited for the networking page. As valuable as this information is, it should not go in the racing tutorial.

CommanderFoo replied:
Going by the outline it was recommended to touch on this subject since it's an advanced scripting tutorial.  So I kept it brief and linked to the networking page which I assume covers more of that subject in detail?  If it doesn't, we should definitely update it so that it contains best practices / example when handling stuff like purchases (not relating to Perks).

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+- Only start races when the server has players.
+- Move players to the starting positions.
+- Let the players know the race is about to begin and when to go.
+- Keep track of each players race time.
```
Keep track of the race time for each player.

(Translation of the original bullet point may not work well in languages that don't convey ownership like English)

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+- Keep track of each players race time.
+- Stop the race when time has ran out before starting a new race.
+
+Create a new script called **Race_Manager_Server** and place it into the **Server Context** folder. Before opening the script we need to setup the custom property for the positions so **Lua** has access to them.
```
Change: 
"place it"
To: 
"place the script you created"

We want to avoid pronouns for translation purposes.

MrMiracleMan111 replied:
Bore opening the script, we need to setup

Added a comma

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+Create a new script called **Race_Manager_Server** and place it into the **Server Context** folder. Before opening the script we need to setup the custom property for the positions so **Lua** has access to them.
+
+!!! tip
+    It's good habit to suffix scripts with the type of context they will be placed in so it's easier to find in the **Project Content** panel. With the addition of **Project Content** folders organizing scripts, templates, and materials is recommended when creating bigger projects.
```
Remove the last sentence: "_With the addition of **Project Content** folders organizing scripts, templates, and materials is recommended when creating bigger projects._"

Replace that sentence with an example case of adding the network context as a suffix to a script name.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+!!! tip
+    It's good habit to suffix scripts with the type of context they will be placed in so it's easier to find in the **Project Content** panel. With the addition of **Project Content** folders organizing scripts, templates, and materials is recommended when creating bigger projects.
+
+We want each player to be moved to a random lane position. At the same time we need to mark each player as in the race to keep track of who is and isn't currently racing because players can join the game while a race is in progress.
```
each player ~~as~~ in the race to keep

remove "as"

MrMiracleMan111 commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4383 @@
+
+```lua linenums="1"
+local function MovePlayersToStart()
+    for k, p in ipairs(Game.GetPlayers()) do
````
Instead of "`k`" and "`p`" use more descriptive names like "`index`" and "`player`" (you could also use "`currentPlayer`")

MrMiracleMan111 commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4383 @@
+players[p.id].inRace = true
+```
+
+We need to keep track of when players are in a race. We can do this by setting a property `inRace` to true that can be used later on.
````
"We can do this by setting a property `inRace` to true that can be used later on."

We can do this by creating a property named `inRace` and setting the property `inRace` to `true`.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4383 @@
+end
+```
+
+The function `OnPlayerJoined` is called each time a player joins the server. To begin with we need to disable player mounts and player jumping. We want this to be a sprint race, so setting `maxJumpCount` to `0` prevents the player from jumping.
````
"To begin with we need to disable player mounts and player jumping."

To begin, we need to disable player mounts and player jumping.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+
+The function `OnPlayerJoined` is called each time a player joins the server. To begin with we need to disable player mounts and player jumping. We want this to be a sprint race, so setting `maxJumpCount` to `0` prevents the player from jumping.
+
+We can also use this handler to push each new player to the `players` table to keep track of players in the game. The value stored in the table is a table with some defaults setup that will be used throughout the script.
```
"We can also use this handler to push each new player"

_Should readers forget what handler you are referring to, restating the name of the handler helps:_

We can also use this `OnPlayerJoined` handler to push each new player

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+
+- `inRace`
+
+    This property gets updated with either `true` or `false`. If a player joins late while a race is already in progress, then we can use this to determine which players are actively racing.
```
"then we can use this to determine which players are actively racing."

*Just restating what we are referring to:*

the new can use this property to determine which players are actively racing.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4383 @@
+end
+```
+
+When a player leaves the game, we need cleanup the `players` table by removing the player from it. Setting the value to `nil` is a good way to tell `Lua` that we no longer need it.
````
"we need cleanup the `players` table by removing the player from it" 

*Small typo and removing the use of "it"*

we need to cleanup the `players` table by removing the player from the `players` table"

MrMiracleMan111 replied:
You can remove the backticks from `Lua`. Lua does not have to be written in code format, it can be written as normal text.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4383 @@
+Game.playerLeftEvent:Connect(OnPlayerLeft)
+```
+
+We need to hook up the events when a player joins and leaves.
````
This sentence may not translate well into other languages.

We need to connect the `OnPlayerJoined` and `OnPlayerLeft` events to the `playerJoinedEvent` event and `playerLeftEvent` event.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+
+### Race Notification
+
+We want to let the players know when to get ready and when to go. Place both functions just after the **OnPlayerLeft** function.
```
Use code format instead of bold for function names

`OnplayerLeft` instead of **OnPlayerLeft**.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4383 @@
+end
+```
+
+The function **TellPlayersToGetReady** will broadcast to each player that is marked as in the race. This is done so players who join late don't receive the notification. Later on in the tutorial we will listen for this broadcast event and setup a handler.
````
Use code format instead of bold for function names

`TellPlayersToGetReady` instead of **TellPlayersToGetReady**.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+
+### Stopping the Race
+
+We need to handle stopping a race when the time is up. So place the function below just after the **TellPlayersToGo** function.
```
The phrase "when the time is up" may not translate well. Instead you could use:

"when the race timer has finished"

MrMiracleMan111 replied:
Make this one sentence:
"We need to handle stopping a race when the time is up. So place the function below just after the **TellPlayersToGo** function."

We need to handle stopping a race when the time is up, so place the function below just after the **TellPlayersToGo** function.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4383 @@
+end
+```
+
+The function **StopRace** will loop through all the players on the server and set the property `inRace` to `false` indicating they are no longer in the race. We then broadcast to all players that the race should be stopped. Later on in the tutorial we will listen for this broadcast event and setup a handler.
````
"Later on in the tutorial we will listen for this broadcast event and setup a handler."

*Small grammatical error. Also, change "setup" to "create" both because "setup" needs to be two words
and because "setup" may not translate well.*

Later on in the tutorial, we will listen for this broadcast event and create a handler.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+
+We need a way to manage the state of the game, and a simple way to handle that for this example is using a task.
+
+![!Flow Chart](../img/RaceTimerTutorial/game_flow_chart.png){: .center loading="lazy" }
```
In the future, try to avoid creating diagrams like these because they will be difficult for non-English speakers to translate.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+
+![!Flow Chart](../img/RaceTimerTutorial/game_flow_chart.png){: .center loading="lazy" }
+
+1. Check if the task has not been setup.
```
This numbered list will need some context beforehand. For example, "Our script will need to be able to complete the following processes:"

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+
+1. Check if the task has not been setup.
+
+    We need to make a repeating task if one hasn't been setup. This task will handle the state of the game by repeating every 10 seconds. Because the length of the track doesn't take very long to finish for the player, we limit the amount of time per race to 10 seconds.
```
"This task will handle the state of the game by repeating every 10 seconds"

Flip this sentence for clarity:

*This task will repeat every 10 seconds and handle the state of the game*

MrMiracleMan111 replied:
"we limit the amount of time per race to 10 seconds."

we limit the duration of each race to 10 seconds.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+
+    We need to make a repeating task if one hasn't been setup. This task will handle the state of the game by repeating every 10 seconds. Because the length of the track doesn't take very long to finish for the player, we limit the amount of time per race to 10 seconds.
+
+    For the initial spawning of the task, we delay it by 14 seconds to take into account for the 4 seconds in the body where `Task.Wait` is used. If we don't do this, then the first race on the server will be short.
```
"For the initial spawning of the task, we delay it by 14 seconds to take into account for the 4 seconds in the body where `Task.Wait` is used."

Clarify what "it" is referring to.

*We delay the initial spawning of the task by 14 seconds to take into account the 4 seconds when `Task.Wait` is used.*

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+
+    For the initial spawning of the task, we delay it by 14 seconds to take into account for the 4 seconds in the body where `Task.Wait` is used. If we don't do this, then the first race on the server will be short.
+
+    Using a repeating task has an added benefit of resetting the race if any players are AFK (Away From KeyBoard).
```
You can replace the "AFK (Away From Keyboard)" section with *inactive*. This helps with translation.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4383 @@
+    raceTask.repeatCount = -1
+    ```
+
+    We want to have the spawned task repeat constantly, so setting `repeatCount` to `-1` allows that.
````
"We want to have the spawned task repeat constantly, so setting `repeatCount` to `-1` allows that." 

We set `repeatCount` to `-1` to force the task to repeat forever.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+
+2. Check if there are players in the game.
+
+    The task handler will check to see if there are any players in the game. If there number of players is `0` then we exit the handler early by using `return`.
```
"If there number of players is `0` then we exit"

*Small typo*

If the number of players is `0`, then we exit

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+
+6. Notify players to go, and enable movement.
+
+    We wait another 2 seconds by using `Task.Wait` and notify the players to go so they know when to start running. At the same time we  enable all players by calling `EnablePlayers`.
```
Double space between "we  enable"

"At the same time we  enable all players by calling `EnablePlayers`."

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+
+### Trigger Overlap
+
+Next we need to modify the **Race_Manager_Server** script so that the finish line trigger can be hooked up to detect when a player has overlapped it.
```
Change "hooked up" to "used".

MrMiracleMan111 commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4383 @@
+
+Next we need to modify the **Race_Manager_Server** script so that the finish line trigger can be hooked up to detect when a player has overlapped it.
+
+Add the following line to the top of the script so that it's under the **STARTING POSITIONS** variable.
```
Use code format for the **STARTING POSITIONS** variable.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4383 @@
+players[obj.id].finishTime = time()
+```
+
+We then set `finishTime` to the current time since the game started. This means that we now have a start and end time which we can use later to determine the players best time.
````
Change:
"players best time"

To: 
best time of the player

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Change:
"players best time"

To:
best time of the player

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Change:
"players best time"

To:
best time of the player

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Change:
"players best time"

To:
best time of the player

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Change:
"players best time"

To:
best time of the player

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Change:
"players best time"

To:
best time of the player

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Change:
"players best time"

To:
best time of the player

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Change:
"players best time"

To:
best time of the player

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Change:
"players best time"

To:
best time of the player

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Change:
"players best time"

To:
best time of the player

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Change:
"players best time"

To:
best time of the player

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Change:
"players best time"

To:
best time of the player

MrMiracleMan111 committed:
Added commas

Added comma after "Next" in the phrase "Next we"

MrMiracleMan111 commented on src/tutorials/race_timer.md:
"true or false"

The words "true" and "false" need to have code formatting.

`true` or `false`

MrMiracleMan111 commented on src/tutorials/race_timer.md:
To follow structure of the previous bullet point, remove the word "this". Also, clarify what the "It" pronoun is referring to. The final message would be something like:

- `timer` will get incremented when the race has stared for the client. The `timer` variable will hold the time for the current race.

MrMiracleMan111 committed:
Added more commas

Added a comma after "Next" where necessary.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Oh, neat tip! I would recommend that instead of giving the reader information about the similarities with C, you should post a link to a website where different string formatting parameters can be found for Lua. I think the website below contains the patterns for `string.format`

[https://www.lua.org/pil/20.2.html](https://www.lua.org/pil/20.2.html)

MrMiracleMan111 committed:
More commas

Added a comma after the word "second" in the sentence section "After 1 second".

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Unless you are referring to all of the players, change "players" to "player's" .

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Unless you are referring to all of the players, change "players" to "player's" .

MrMiracleMan111 committed:
Small grammatical fixes.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
I would recommend changing "dealing with" to "using" to improve the translation.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
change "it" to "the script"

MrMiracleMan111 commented on src/tutorials/race_timer.md:
In the section "It will update the leaderboards" replace "It" with "The task".

MrMiracleMan111 commented on src/tutorials/race_timer.md:
In the section "It fetches each entry", replace "It" with "The code"

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Specify which script line 10 is referring to.

CommanderFoo replied:
I was a little bit on the fence with specifying the script, as the section header states which script it is they are modifying.  However, I can see how easily it could be for the user to forget which script it is they are modifying and needing to scroll back up to see the header.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Specify which script line 8 is referring to.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Specify which script line 12 is referring to.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Specify which script line 5 is referring to.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Specify which script line 11 is referring to.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Specify which script line 11 is referring to.

MrMiracleMan111 commented on src/tutorials/race_timer.md:
Replace "it" with "the script"

MrMiracleMan111 commented on src/tutorials/race_timer.md:
"This function" -> state the name of the function in Code format

Replace "It" with the name of the function.

MrMiracleMan111 reviewed (changes requested):
The tutorial was very informative (I even learned a few things). You did a fantastic job thoroughly explaining how code worked and how players put the pieces together. Sadly the tutorial may be too long. To shorten this tutorial you could add some steps into the Community Content portions. For example, you could create a folder of section split triggers so that you don't have to explain that portion as much.

Other than length, keeping track of sentences that could be mistranslated. For example, the phrase "time is up" is common in English, however, it may have no meaning or a different meaning in other languages.

 Also, try to avoid using pronouns like "it" and instead state the subject of your sentence, which helps with translation.

PopThosePringles committed:
Updated changes

Made changes based on conversation.
Flow chart removed.  Was not needed anyway.

StanzillaManticore commented on mkdocs.yml:
```diff
@@ -60,6 +60,7 @@ nav:
               - Scripting (Beginner): tutorials/lua_basics_helloworld.md
               - Scripting (Intermediate): tutorials/lua_basics_lightbulb.md
               - Scripting (Advanced): tutorials/lua_basics_manticoin.md
+              - Scripting Race Timer: tutorials/race_timer.md
```
should we call it `Scripting (Expert)`?

CommanderFoo replied:
Yeah, I wasn't sure on this one. If we follow the pattern of the previous ones, then yes, that sounds good.

If we do one that's even more advanced than "Expert", then we might need to come up with something else 😛

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+* **Skills you will learn:**
+    * When to use Client and Server contexts
+    * Accurate time keeping
+    * Broadcasting to the client
```
Communicating from server to client

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+![!Remove Default Floor](../img/RaceTimerTutorial/remove_default_floor.png){: .center loading="lazy" }
+
+We now need to create a track for players to run on. A track has been uploaded to **Community Content** to help you get started. This can be imported from the **Community Content** panel by searching for **Race Timer Tutorial - Track** by **CommanderFoo**.
```
We should republish on CoreAcademy account

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+![!Community Content Template](../img/RaceTimerTutorial/import_track.png){: .center loading="lazy" }
+
+Drag and drop the template **Race Timer Tutorial - Track** into the **Hierarchy** and **Deinstance** it.
```
right click and select **Deinstance This Template**

Essentially, we need to decide if the readership is starting here (possible, if they have some experience elsewhere but want to learn Core) or if we're going to assume they've done some other tutorials. If the former, we should explain the mechanics of deleting and deinstancing explicitly, and if the latter, we probably don't need the images.

I'm not sure which is correct here, but I'm leaning towards the second scenario. Most of the demand for this is from creators who have already done Lightbulb, and advanced creators have the option of looking back if they can't figure out a step.

CommanderFoo replied:
I agree, the second scenario is better, since this is aimed at more experienced scripters. 

I've cut out some of the beginner type stuff in the early section, and removed a few images and videos.

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+## Starting Line
+
+When a race is about to start, we need to move the players to the starting line. Instead of grouping all players up in one spot, we can create a few positions along the starting line and randomly pick one to place the player at. There is still a possibility of players spawning in the same location, and to fix this it would require a more advanced system to keep track of which lanes have been assi ...
```
Move the second sentence into a !!! note

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+When a race is about to start, we need to move the players to the starting line. Instead of grouping all players up in one spot, we can create a few positions along the starting line and randomly pick one to place the player at. There is still a possibility of players spawning in the same location, and to fix this it would require a more advanced system to keep track of which lanes have been assi ...
+
+### Creating Lane Positions
```
Create

PopThosePringles committed:
Improved tips

Some tips are now inline (no title).
Some now have titles for them.
Changed nav name.

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+Understanding the difference between these two contexts can help you understand where to place **Lua** scripts and what they should be used for.
+
+Anything in the **Client Context** exists on the clients version of the game. This means that there is no network traffic, and anything that gets modified is applied to that local player and not all players on the server.
```
* on each player's version of the game.

Also, just a side note here, I keep finding client context things that are surprisingly silently replicated anyway. Is this just me? IK anchors, apparently client-side trigger interaction events (only).

CommanderFoo replied:
Hmm, that's interesting about client things being replicated. Maybe I should change how I phrase it so say "most things"?

"Most things in the **Client Context** exists on each player's version of the game."

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+Anything in the **Client Context** exists on the clients version of the game. This means that there is no network traffic, and anything that gets modified is applied to that local player and not all players on the server.
+
+With **Lua** scripts, nothing sensitive should be stored in a **Client Context** script, and any action the client can do should always be sent to a script in a **Server Context** to validate it. **The client should never be trusted.**
```
"trusted" may be a little vague here... The client should never be treated as an authority on the state of the game for all players?

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+    </video>
+</div>
+
+## Setup Script Hierarchy
```
Rename this section Client and Server context, and creating another h3 with this title just under it. 

Setup --> Set up

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+## Setup Script Hierarchy
+
+We need to setup the script folder and contexts in the **Hierarchy**.
```
Try to stick with "you" as the subject unless it's particularly awkward.

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+2. Create a **Client Context** inside the **Scripts** folder.
+3. Create a **Server Context** inside the **Scripts** folder.
+
+### Client Context vs Server Context
```
Need a more specific title for this section. Server Authoritative Design is close ... something that tackles that this not explaining hte difference between these contexts but showing how to organize code to avoid hacking.

coreslinkous commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4385 @@
+end)
+```
+
+In the above script, see line 9 where the price and the ID of the shop item is being sent to the server. This is bad because the broadcast to the server is also sending the price of the item which could be modified by the client.
````
Try to condense into one sentence anywhere where the first sentence is just asking the reader to notice. Skip straight to what it does.

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+!!! tip
+    There is another context called **Default Context**. Any script in this context is also viewable by the client. So if you have code you want to protect from prying eyes, consider moving it to a **Server Context**.
+
+For further information checkout [Networking in Core](../references/networking/)
```
checkout --> check out

coreslinkous commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4385 @@
+end)
+```
+
+In the above script, see line 9 where the item ID is sent to the server only. The server script should then handle validating if the player can purchase the item, and if so it will initiate the purchase and possibly broadcast back to the client if it was successful or not.
````
same comment as above

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+For further information checkout [Networking in Core](../references/networking/)
+
+## Race Manager Server Script
```
Start h2 sections with a gerund ... "Creating the Race Manager Server Script" or building/designing/writing

PopThosePringles committed:
Made changes based on review comments

Removed some of the "beginner" parts of the tutorial from the beginning to help move things along quicker.

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+- Keep track of the race time for each player.
+- Stop the race when time has ran out before starting a new race.
+
+Create a new script called **Race_Manager_Server** and place the script you created into the **Server Context** folder. Before opening the script, we need to setup the custom property for the positions so **Lua** has access to them.
```
Should discuss convention here. Manticore has tended to do PascalCase, but what I've seen from Meta underscores to separate the context ``RaceManagerServer`` or ``RaceManager_Server``

StanzillaManticore replied:
`RaceManager_Server`

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+Open up the **Race_Manager_Server** script by double clicking on it.
+
+!!! tip
+    **Core** comes with a script editor built in. You can change which editor is used in the **Settings** under the **Editor** section. Another popular editor is **Visual Studio Code** which has an extension for the **Core API**.
```
- Link to page that explains setting up external editor.
- Use the official window name **Script Editor**
- Probably should mention the advantages of the Script Editor casually here, like the fact that you can nest it like any window, and the F5 compile. Probably don't need to go as far as breakpoints, but that's cool too!

coreslinkous commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4385 @@
+local players = {}
+```
+
+- `START_POSITIONS`
````
ul or table

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+    We need to keep track of the players in the game. A good way to do that is putting players that have joined into a table for access later on.
+
+!!! tip
+    The word **Handler** is just a more specific term for **function** that is used with events. **Handlers** are also known as **Callbacks**.
```
Do Lua scripters use the term Callback? I honestly don't know, but callbacks as I understand them are tricky to implement in Lua.

tip --> info

StanzillaManticore replied:
Yeah callback is common

coreslinkous commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4385 @@
+end
+```
+
+The code above contains two helper functions that will allow us to disable and enable player movement. We need to prevent players who are at the starting positions from moving.
````
"you" subject not "we"

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+!!! tip
+    Any function that doesn't need to be accessed outside of the script should be marked as `local`. This doesn't include functions that are overridden like `Tick`.
+
+### Move Player to Starting Line
```
So, if we are assuming that the *only* preparation the reader has had before this is lightbulb, I'm pretty sure we haven't talked about iteration, so it's worth a section to just quickly explain iteration using ipairs. I think you can confine to the numbered variety now, but a graphic would be great here showing a list of players each becoming player as the loop repeats. We could gather the keyframes and give them to Stephano.

CommanderFoo replied:
Will take a look at this one tomrorow and see what I can do.

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+We can also use the handler `OnPlayerJoined` to push each new player to the `players` table to keep track of players in the game. The value stored in the table is a table with some defaults setup that will be used throughout the script.
+
+- `inRace`
```
ul or table

coreslinkous commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4385 @@
+Game.playerLeftEvent:Connect(OnPlayerLeft)
+```
+
+We need to connect the `OnPlayerJoined` and `OnPlayerLeft` events to the `playerJoinedEvent` event and `playerLeftEvent` event.
````
"You" subject, or reframe it to have the code be the subject to be more direct. It *does* that, it's not still a problem to solve.

coreslinkous commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4385 @@
+end
+```
+
+The function `TellPlayersToGetReady` will broadcast to each player that is marked as in the race. This is done so players who join late don't receive the notification. Later on in the tutorial we will listen for this broadcast event and setup a handler.
````
It broadcasts only to those specific players, right? Worth pointing out that those clients will get the message and others will not.

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+We need to keep track of when players are in a race. We can do this by setting the property `inRace` to `true`.
+
+### Enabling Players in Race
```
Command form. Allow Players to Race?

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+!!! tip
+    `time()` returns the time in seconds (floating point) since the game started on the server.
+
+### Player Joined / Left Events
```
Command form. Just think like, if this were a series of numbered steps with no further explanation, each h3-level heading should look like one of those steps.

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+We need to connect the `OnPlayerJoined` and `OnPlayerLeft` events to the `playerJoinedEvent` event and `playerLeftEvent` event.
+
+### Race Notification
```
Command form

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+    </video>
+</div>
+
+Enter **Play** mode and test everything is working. If you have followed the tutorial so far, then this is what happens:
```
Separate h3 for the testing step.

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+3. After another few seconds the player can now move.
+4. Every 10 seconds the task will reset the player back to a random starting position.
+
+??? "Full **Race_Manager_Server** Script"
```
h3. We want the impatient to be able to find it from the ToC, but it is still part of this section.

coreslinkous replied:
And because this is a different kind of information, we can break the h3's-should-be-in-imperative-form rule

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+2. Create Trigger
+
+    Create a new **Trigger** inside the **Triggers** group and name it **Finish**.
```
I use "add" when it is pulled from Core Content, and "create" if you can do it with the right click menu in the Hierarchy

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+Next, we need to modify the **Race_Manager_Server** script so that the finish line trigger can be used to detect when a player has overlapped it.
+
+Add the following line to the top of the script so that it's under the `STARTING POSITIONS` variable.
```
``START_POSITIONS``?

Also you can skip things like "so that" and just directly say where it should be.

coreslinkous commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4385 @@
+end
+```
+
+The trigger handler above will make sure that the `obj` is valid and is a `Player`. We then check to see if the player is in the `players` table and if `inRace` is marked as true. We do this so that players who join the server while a race is in progress don't get detected when they overlap the trigger.
````
and is a Player type of object? Make it clear that this is a type that you can check for so it extends to other uses later.

coreslinkous commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4385 @@
+players[obj.id].inRace = false
+```
+
+At this point, the player is now marked as not in the race. This prevents the overlap handler from triggering again.
````
"This point" being when the race has finished for them individually.

coreslinkous commented on src/tutorials/race_timer.md:
````diff
@@ -0,0 +1,4385 @@
+print(finalTime)
+```
+
+ To test that the time is being recorded correctly, we can print out `finalTime` to the **Event Log**.
````
Can this be broken into more steps? It is getting long.

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+Enter **Play** mode and test everything is working. If you have followed the tutorial so far, then this is what happens:
+
+1. Player spawns in at the **Spawn Point**.
```
The player

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+Enter **Play** mode and test everything is working. If you have followed the tutorial so far, then this is what happens:
+
+1. Player spawns in at the **Spawn Point**.
+2. After a few seconds the player movement is disabled and moved to a random starting position.
```
the player's movement

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -0,0 +1,4385 @@
+
+3. Add Custom Property
+
+    The **Race_Manager_Server** script needs to know about the finish line trigger. Click on the **Race_Manager_Server** script so it becomes the active object in the **Hierarchy** and drag and drop the **Finish** trigger onto the **Add Custom Property** button. Rename the new custom property to **finishTrigger**.
```
"know about" might be vague. "Have a way to reference/talk about"

drag and drop -> drag

coreslinkous commented on src/tutorials/race_timer.md:
Passive voice: Client can't be trusted

coreslinkous commented on src/tutorials/race_timer.md:
setup -> set up (always spaced if it's a verb)

Also, something more specific than "a few different". The following?

coreslinkous commented on src/tutorials/race_timer.md:
Don't need to say that it's later. We can compartmentalize creating all the UI spaces from a design perspective and expect to worry about some of the implementation later.

coreslinkous commented on src/tutorials/race_timer.md:
All UI should

coreslinkous commented on src/tutorials/race_timer.md:
As a child?

coreslinkous commented on src/tutorials/race_timer.md:
Rework to not have images in between the ol steps. Probably fewer images, one before and after the ol.

coreslinkous commented on src/tutorials/race_timer.md:
command form

coreslinkous commented on src/tutorials/race_timer.md:
Command form

coreslinkous commented on src/tutorials/race_timer.md:
Adding and custom property adding can be its own h3 step

coreslinkous commented on src/tutorials/race_timer.md:
And then this is a new h3 step

coreslinkous commented on src/tutorials/race_timer.md:
Need a warning in this section that if you are referencing any part of the UI besides the UIText, it won't have a 'text' property and you will get the error ... [insert precise language from the editor here].

coreslinkous commented on src/tutorials/race_timer.md:
Just link directly to the Lua docs page instead.

CommanderFoo replied:
Unfortunately the Lua docs doesn't contain the list of options or modifiers. Ideally we could do with having one on docs with all the options and modifiers. For now I will link to the Lua docs that mentions a little bit about string.format. I always find myself forgetting some of the options, and go searching for the list.

coreslinkous commented on src/tutorials/race_timer.md:
- cut "to the player"
- So we --> To do this,
- cut "it's visual only"

The other advantage of doing it on the client is that it can refresh quickly and updates smoothly

coreslinkous commented on src/tutorials/race_timer.md:
Need to figure out how to break this section into more steps as well.

coreslinkous replied:
Maybe even one h4 per code block?

This would also solve the problem that everyone sees the code before learning what it does. A title could hint it, let you investigate for yourself, and then fully explain.

coreslinkous commented on src/tutorials/race_timer.md:
I would do the test step first, and give specific instructions for how to verify that each part is working. Then the final final part is the complete code.

coreslinkous commented on src/tutorials/race_timer.md:
Cut "it would be nice". This is what you're doing next. 
player's
Cut "here is an overview". These are the steps to do X

coreslinkous commented on src/tutorials/race_timer.md:
Change this to On the Server, or frame around that this is the server section

OR rewrite to be a list of script responsibilities, as suggested in the earlier section overview

coreslinkous commented on src/tutorials/race_timer.md:
In general, think about writing this like a spec, where a creator who is feeling confident or more experienced could just take it from here and try it themselves

coreslinkous commented on src/tutorials/race_timer.md:
Briefly explain what player storage actually is.

May want to add a tip that you can add Game Settings from Core Content if you have deleted it, but you don't want two. Probably don't need the tip that it will throw a warning because, well, they will discover that themselves.

coreslinkous commented on src/tutorials/race_timer.md:
New h3 step

coreslinkous commented on src/tutorials/race_timer.md:
h3 again, and frame this one as a noun. The Updated Race_Manager_Server Script

coreslinkous commented on src/tutorials/race_timer.md:
Gerund. "Updating the X UI"

coreslinkous commented on src/tutorials/race_timer.md:
Try to remove images between numbers in an ol

coreslinkous replied:
You could convert each number to an h4 instead, but this does seem like a lot of images

PopThosePringles committed:
Changes based on review recommendations.

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -1190,33 +1230,28 @@ Finally the functions are connected to the events which are called from the serv
-!!! tip "Any script that attempts to access or save to player storage will throw a warning in the **Event Log** letting you know that it isn't enabled."
+!!! tip "If you are missing the **Game Settings** object from your **Hierarchy**, you can add it back in from **Core Content**."
 
 ### Sending Data to Client
```
Sending -> Send

coreslinkous commented on src/tutorials/race_timer.md:
```diff
@@ -1190,33 +1230,28 @@ Finally the functions are connected to the events which are called from the serv
+    - Wait for data to be received
+    - Update UI with fastest time
+
+### Player Storage
```
Enable Player Storage

PopThosePringles committed:
Changes based on review

PopThosePringles committed:
Missing quotes

PopThosePringles committed:
Changed CC name to CoreAcademy

PopThosePringles committed:
Changed some titles

PopThosePringles committed:
Changed we to you, and would to will
